### PR TITLE
fix(core): stop a latched blocked status from outliving a finished turn

### PR DESCRIPTION
**Pull request by LeTuR**
## Intent

Captain's words, 2026-09-06, on seeing a live session in his interface:

"Current status is showing 'blocked' for thurbox frozen bug fix related session. Is it expected behavior ? it should show idle/done instead ?"

and then, authorizing the fix:

"Fix the blocked bug if possible"

No, it is not expected. It should read idle or done. A session whose agent has finished must not read 'blocked'. He approves every merge himself - do not merge.

Later words from the captain, on the running-state glyph in the same interface: he can see the new running glyph, but 'it is a bit ugly and do not render nicely in the UI', and he asked why it is not the animated glyph. He then approved the recommendation put to him with 'Go with recommended way !' - that recommendation was to replace the running glyph with a better-supported codepoint, and to animate the running state only when the pane is actually producing output (reusing the existing spinner), so the indicator never claims knowledge it lacks.

Later still, asked whether that animation should be built now or as a separate change, the captain said: 'I want greptile confident to be 5/5'. The greptile review holds this PR at 4/5 and calls it not yet safe to merge for exactly one outstanding reason, with nothing else on its list: a running session whose pane is producing output still shows a static indicator instead of the approved output-gated spinner. So the animation is to land in this PR rather than as a follow-up.

## What Changed

- Fixed the root cause of a session showing `blocked` forever after its turn ended: the Claude/Antigravity/Grok `Notification` hook commands matched permission/approval keywords against the **entire** payload (including `cwd`/`transcript_path`), so a checkout path containing a word like "permission" made the idle notification that fires after every turn falsely signal `blocked` — with no later hook to overwrite it. The commands now extract and match only the payload's `message` field (`extensions/hooks/claude.json`, `antigravity-hooks.json`, `grok-hooks.json`, bumped to hooks extension v1.9.0).
- Extended `with_output_quiescence` with an `outlived_by_output` check that folds a `blocked` state back to `working`/`idle` once its pane keeps printing well past the state's stamped age, and fixed the reader loop (`src/agent/backend.rs`) to stop counting an adopted session's replayed tmux scrollback as fresh output (`AdoptedSession::seed_len`), which previously made a real hours-long block look instantly quiet and get folded away.
- Added `Coverage::Presumed` / `CoverageSource::ByDetection` so a session whose own agent name resolves to no coverage, but whose pane is observed running a recognized agent, is diagnosed and reported accordingly (`src/session/hook_status.rs`, `src/cli/session_doctor.rs`).
- Replaced the static `running` glyph with a more widely-supported codepoint and made it animate (reusing the existing working spinner) only while `thurbox.printing` reports the pane is actually producing output, so the indicator never claims a turn is in flight when it can't tell (`ui/lib/theme.lua`, `ui/lib/ui.lua`, `ui/plugins/10_sessions.lua`).
- Plumbed a `printing` set/version end-to-end (`Terminals::sync_printing`/`printing_version`, `Epoch::printing`, `Published::printing`, published as `thurbox.printing`) and gated the render-loop animation clock on it so a printing `running` session keeps ticking (`src/kernel/host/mod.rs`, `src/kernel/host/publish.rs`, `src/kernel/snapshot.rs`, `src/coordinator/mod.rs`, `src/coordinator/publish.rs`).
- Updated docs and skills (`docs/AGENTS.md`, `docs/FEATURES.md`, `docs/PERFORMANCE.md`, `docs/ORCHESTRATION.md`, `.agents/skills/thurbox-session-status/SKILL.md`, `.agents/skills/thurbox-cli/SKILL.md`, `extensions/hooks/README.md`) and test suites (`src/session_ops/builtin_hooks.rs`, `tests/kernel_mvp.rs`, `tests/kernel_frame_cost.rs`, plus per-test `thurbox.printing` publish updates across `tests/*.rs`) to match.

## Design note: where the animation had to hook in

The greptile review named `ui/lib/ui.lua`'s `ui.status` as the place `running`
stayed static. That is where the *symptom* shows, but the blocker sat one layer
down, and it is worth recording because it decided the shape of this change.

`ui.status` could not animate `running` on evidence because no evidence reached
Lua: nothing per-row said whether a pane was printing. `src/kernel/host/publish.rs`
gates the `sessions` group on `[epoch.snapshot, epoch.meta, epoch.failed, 0]` and
**deliberately not** on the animation clock — the source calls it "the largest
group by a distance", a table per session with ~30 named fields. Hanging a
printing flag on those rows would have rebuilt all of them every time any pane
started or stopped producing output, which is the cost that gating exists to
prevent (ADR-P16).

So the fact travels in a group of its own: `thurbox.printing`, one session id per
printing session, gated on `epoch.printing` /
`Terminals::printing_version`. That version counts **set membership**, not the
output clock — the clock moves on every byte, and a group gated on it would
rebuild on every frame under a working agent and be worth nothing.
`tests/kernel_frame_cost.rs` pins that a printing change rebuilds exactly one
group, counted against a control rather than a literal so it survives the next
group anyone adds.

Two consequences fall out of that seam rather than being chosen separately.
Only a surface holding the terminals can fill the set, so a headless reader
publishes none, `ui.printing` answers false, and the glyph is static — the same
line `session get` already draws at the folds it cannot make. And
`advance_animation` had to learn the new case, or the clock it gates would stop
and freeze the spinner mid-turn.

## Risk Assessment

✅ Low: The two reviewed commits (9e495b5b: output-gated running-state animation; b730de33: reader_loop seed_len fix for the blocked-latch scrollback-replay bug) are well-scoped, match established architectural patterns (single-signal published groups, change-signal-not-value-signal gating), are covered by targeted tests including a regression test reproducing the original blocked-latch bug, and satisfy the authoritative user intent (running glyph now animates only on output evidence, addressing the sole outstanding greptile concern). Traced the seed_len byte-accounting against std::io::Chain's actual read semantics (it never mixes bytes from two sources within one read() call) and confirmed the fix correctly gates last_output_at stamping on live vs. replayed bytes in all boundary cases.

## Testing

Baseline `cargo nextest run --all` already passed per the pipeline; on top of that I ran the 54 tests directly targeting this change's two fixes (the blocked-latch output-quiescence fold, the scrollback-replay guard on last_output_at, and the running-glyph output-gated animation), all passing, and additionally produced a rendered-TUI text artifact (via a temporary, since-reverted test) showing the running indicator drawing as a static ◉ glyph while the pane is quiet and as an animated spinner frame while the pane is printing — directly demonstrating the exact behavior the greptile review was blocking on. The working tree is clean (only evidence file written outside it).

<details>
<summary>Evidence: Rendered TUI frames: running-status glyph static vs. animated on pane output</summary>

```text
=== running, pane quiet (static glyph, no spinner) ===
╭ Sessions ───────────────────────────────────◉╮
│── thurbox ───────────────────────────────────│
│ ◉ ⑂ fm-worker  agent running · no status rep…│
│                                              │
│                                              │
│                                              │
│                                              │
│                                              │
│                                              │
│                                              │
│                                              │
╰──────────────────────────────────────────────╯

=== running, pane printing (animated spinner) ===
╭ Sessions ───────────────────────────────────⠇╮
│── thurbox ───────────────────────────────────│
│ ⠇ ⑂ fm-worker  agent running · no status rep…│
│                                              │
│                                              │
│                                              │
│                                              │
│                                              │
│                                              │
│                                              │
│                                              │
╰──────────────────────────────────────────────╯
```
</details>
- Outcome: 🔧 1 issue found → auto-fixed ✅ across 2 runs (6m7s)

## Pipeline

Updates from [git push no-mistakes](https://github.com/kunchenguid/no-mistakes)

<details>
<summary>✅ **intent** - passed</summary>

✅ No issues found.
</details>

<details>
<summary>✅ **Rebase** - passed</summary>

✅ No issues found.
</details>

<details>
<summary>🔧 **Review** - 1 issue found → auto-fixed ✅</summary>

- 🚨 `src/session/hook_status.rs:902` - `outlived_by_output` (and its caller `apply_output_quiescence` in src/kernel/snapshot.rs:996) trusts `quiet_for_ms` (derived from `Terminals::millis_since_output`, ultimately `WiredPane::last_output_at`) as evidence that a pane is genuinely producing fresh output, and compares it against `state_age_ms` (the DB&#39;s `hook_state_at`, an absolute historical timestamp) to decide a standing `blocked` is stale. But `last_output_at` is stamped on *every* successful read from the pane&#39;s reader thread (src/agent/backend.rs:995), and that reader is fed `Cursor::new(seed).chain(connected.output)` (src/agent/tmux.rs:1600-1605) — i.e. on every session *adopt* (which is exactly what happens for every already-running session across a thurbox restart, crash+respawn, or first-attach), the captured tmux scrollback (`capture_history_seed`) is replayed through the same reader and immediately bumps `last_output_at` to &#39;now&#39;, indistinguishable from real live output. `initial_output_at(WireMode::Adopt)` deliberately starts at 0 specifically to avoid a post-adopt repaint reading as activity (see the `wire_mode_adopt_starts_stale_spawn_starts_fresh` test/comment in backend.rs), but that protection is defeated the instant the seed bytes are read, which is normally the very first read the reader thread performs.

Concrete failing sequence: a session enters `blocked` at t0 (hook_state_at = t0, e.g. a real permission prompt) and stays genuinely blocked for hours with no further pane output. Thurbox restarts (or the session panel is (re)attached) at t1 = t0 + 3h. `Terminals::sync` adopts the pane, replaying scrollback; `last_output_at` becomes ~t1. On the very next tick, `apply_output_quiescence` computes `age = t1 - t0` (~3h) and `quiet ≈ 0`, so `age.saturating_sub(quiet) &gt; WORKING_QUIET_MS` is true, and `with_output_quiescence` folds `Blocked` into `Working`/`Idle`. Because nothing ever re-stamps `hook_state_at` (no new hook fires) and nothing further prints (the session really is waiting on the user), `age` and `quiet` grow at the same rate afterward, so the gap never closes — the misclassification is permanent, not a transient blip like the analogous `working`→`idle` case (which is purely window-based and self-corrects). The session silently and durably stops showing `blocked` even though the user still needs to respond — the same &#39;agent needs my attention but the UI won&#39;t show it&#39; failure class this PR set out to fix, just reached from the opposite direction and, unlike the original bug, invisible until someone happens to check the pane directly.

Earliest shared boundary to fix this: distinguish genuine live output from scrollback replay at the one place both consumers of `last_output_at` (`sync_printing`&#39;s animation gate and `outlived_by_output`&#39;s blocked fold) rely on it — e.g. don&#39;t stamp `last_output_at` for bytes read from the `Cursor::new(seed)` half of the chained reader in src/agent/tmux.rs / src/agent/backend.rs&#39;s `reader_loop`, only for bytes from the live `connected.output` stream.

🔧 Fix: wait for verification: agent:: test suite running in background
✅ Re-checked - no issues remain.
</details>

<details>
<summary>🔧 **Test** - 1 issue found → auto-fixed ✅</summary>

- 🚨 tests failed with exit code 101
- `cargo nextest run --all`

🔧 Fix: fix: clear stale incremental build artifacts causing link failure
✅ Re-checked - no issues remain.
- `cargo nextest run --all`
- `cargo nextest run -E 'test(hook_status) or test(apply_output_quiescence) or test(reader_loop) or test(outlived) or test(a_block_the_pane) or test(coverage_falls_back) or test(detected_coverage) or test(a_declared_agent_keeps) or test(an_unnamed_foreign_agent) or test(a_standing_block) or test(scrollback)' — 44 passed`
- `cargo nextest run -E 'test(a_running_session_animates_only_while_its_pane_prints) or test(output_animates_running_and_nothing_else) or test(a_working_session_animates_without_needing_the_pane) or test(a_change_of_what_is_printing_rebuilds_one_group_and_no_others) or test(a_block_the_pane_printed_past_stops_being_drawn) or test(hooks_expected) or test(a_partial_agent_warns_without_failing) or test(an_agent_with_no_wiring_fails_and_says_how_to_wire_it) or test(a_session_with_no_reported_status_never_draws_the_idle_dot) or test(a_running_session_with_no_determined_agent_says_so_without_a_name)' — 10 passed`
- <code>manual: appended a temporary #[test] to tests/kernel_mvp.rs that renders the actual ratatui TUI frame for a `running` session via publish_printing()+paint() with the pane quiet vs. printing, wrote both rendered frames to the evidence file, ran it, then reverted the temporary test with `git checkout -- tests/kernel_mvp.rs` (confirmed clean working tree afterward)</code>
</details>

<details>
<summary>✅ **Document** - passed</summary>

✅ No issues found.
</details>

<details>
<summary>🔧 **Lint** - 1 issue found → auto-fixed (2) ✅</summary>

- ⚠️ linter found issues (exit code 1)

🔧 Fix: {&#34;summary&#34;: &#34;style: fmt reader_loop and assert_eq call sites in backend.rs&#34;}
1 warning still open:

- ⚠️ linter found issues (exit code 101)

🔧 Fix: docs: replace private intra-doc link with plain code span
✅ Re-checked - no issues remain.
</details>

<details>
<summary>✅ **Push** - passed</summary>

✅ No issues found.
</details>

### CODE REVIEW

File: ui/lib/theme.lua [hunk 1]
```diff
@@ -99,7 +99,12 @@ local STATUS_GLYPHS = {
   -- An agent holds the pane and has reported nothing. Filled, because
   -- something IS there; not the working spinner, because no process listing
   -- can tell a turn in flight from a prompt waiting for input.
-  running = "◍",
+  --
+  -- FISHEYE rather than CIRCLE WITH VERTICAL FILL (U+25CD), which is rare
+  -- enough that monospace fonts routinely have no glyph for it and draw a
+  -- fallback box. Still distinct from done's solid circle and idle's hollow
+  -- one, which is the whole job of this row.
+  running = "◉",
```
[the remark this answers is not in the document]
Comment (LeTuR) [reply ui/lib/theme.lua:109]: Correct, and deliberate — see **Scoped Out: animating the running state** in the PR description.

The blocker is not `ui.status`; it is that nothing tells the Lua whether a pane is printing. `src/kernel/host/publish.rs` gates the `sessions` group on `[epoch.snapshot, epoch.meta, epoch.failed, 0]` and deliberately *not* on `epoch.animation`, calling it "the largest group by a distance". Carrying a per-row printing fact means either adding `epoch.animation` to that gate — rebuilding the largest published table at `ANIMATION_HZ` for every session whenever anything animates — or adding a new animation-gated publish group plus its `thurbox.d.lua`/`thurbox.yml` surface, which is a new public surface for plugins.

That is a render-loop change-gate decision, so it was split into its own change rather than bolted onto this status bugfix. Nothing half-built is left behind: `running` stays static, and `docs/FEATURES.md` and the `thurbox-session-status` skill document it as static, so no docs pre-announce motion that does not exist.

[the remark this answers is not in the document]
Comment (LeTuR) [reply ui/lib/theme.lua:109]: **Correction — this now ships in this PR.** My earlier reply said the animation was deliberately deferred; the captain reversed that call, so it landed here rather than as a follow-up.

`running` now animates through the existing `theme.spinner` while its pane is producing output, and falls back to the static fisheye the moment it goes quiet. Your diagnosis was right that `ui.status` only animated `working` — the reason it could not do better was that no evidence reached Lua. That is fixed at the seam: `thurbox.printing` is a new publish group (one session id per printing session, gated on `epoch.printing` / `Terminals::printing_version`, which counts set membership rather than the output clock), kept separate from the `sessions` group so the largest published table is not rebuilt every time a pane starts or stops printing.

`tests/kernel_mvp.rs` covers both arms — a printing pane animates, a quiet one does not, and `uncovered`/`unreported`/`idle` stay still even when listed as printing — and `tests/kernel_frame_cost.rs` pins that a printing change rebuilds exactly one group.

One thing your review indirectly caught: the same output signal was also feeding the new blocked-latch fold, and it was corrupted by scrollback replay on session adopt, which would have silently retired a genuinely blocked session after any restart. Fixed in the same PR (`AdoptedSession::seed_len`), with a test that a three-hour-old `blocked` survives replay.
